File: app/src/test/java/openfoodfacts/github/scrachx/openfood/category/CategoryRepositoryTest.java
package org.openfoodfacts.scanner.category;

import org.hamcrest.CoreMatchers;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import io.reactivex.Single;
import io.reactivex.observers.TestObserver;
import org.openfoodfacts.scanner.category.mapper.CategoryMapper;
import org.openfoodfacts.scanner.category.model.Category;
import org.openfoodfacts.scanner.category.network.CategoryNetworkService;
import org.openfoodfacts.scanner.category.network.CategoryResponse;

/**
 * Created by Abdelali Eramli on 01/01/2018.
 */

@RunWith(MockitoJUnitRunner.class)
public class CategoryRepositoryTest {
    @Mock
    private CategoryMapper mapper;
    @Mock
    private CategoryNetworkService networkService;
    @Mock
    private Category category;
    @Mock
    private CategoryResponse response;

    private CategoryRepository repository;

    @Before
    public void setup() {
        Mockito.when(mapper.fromNetwork(ArgumentMatchers.any())).thenReturn(Arrays.asList(category, category, category));
        Mockito.when(networkService.getCategories()).thenReturn(Single.just(response));
        repository = new CategoryRepository(networkService, mapper);
    }

    @Test
    public void retrieveAll_Success() {
        TestObserver<List<Category>> testObserver = new TestObserver<>();
        repository.retrieveAll().subscribe(testObserver);

        testObserver.awaitTerminalEvent();
        List<Category> result = testObserver.values().get(0);
        Assert.assertThat(result.get(0), CoreMatchers.is(category));
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/category/CategoryRepositoryTest.kt
package openfoodfacts.github.scrachx.openfood.category

import com.google.common.truth.Truth.assertThat
import io.mockk.coEvery
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import openfoodfacts.github.scrachx.openfood.category.mapper.CategoryMapper
import openfoodfacts.github.scrachx.openfood.category.model.Category
import openfoodfacts.github.scrachx.openfood.category.network.CategoryNetworkService
import openfoodfacts.github.scrachx.openfood.category.network.CategoryResponse
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Created by Abdelali Eramli on 01/01/2018.
 */
@OptIn(ExperimentalCoroutinesApi::class)
@ExtendWith(MockKExtension::class)
class CategoryRepositoryTest {
    @MockK
    private lateinit var mapper: CategoryMapper

    @MockK
    private lateinit var networkService: CategoryNetworkService

    @MockK
    private lateinit var category: Category

    private val response = CategoryResponse()
    private lateinit var repository: CategoryRepository

    @BeforeEach
    fun setup() {
        every { mapper.fromNetwork(any()) } returns listOf(category, category, category)
        coEvery { networkService.getCategories() } returns response

        repository = CategoryRepository(networkService, mapper)
    }

    @Test
    fun retrieveAll_Success() = runTest {
        val result = repository.retrieveAll()!![0]
        assertThat(result).isEqualTo(category)
    }

}